Prompt: avoid resetting PR/issue metadata

cristiand391 opened the pull request:
Fixes #2369 
Fixes #2475 

The current implementation of `MetadataSurvey` resets metadata provided by flags if no option is selected (press `Enter` without selecting anything) and when the user adds metadata in the prompt.
This happens because Survey doesn't return an error if you pass an empty slice of questions to it, so [this check pass](https://github.com/cli/cli/blob/9f84f0ffa1d5b3141399b7ed499eb81b69ef74d2/pkg/cmd/pr/shared/survey.go#L349) and `state` metadata fields are assigned to `nil` after that.

This PR contains the following changes:

- Check if `mqs` is empty before sending it to survey
- If `mqs` isn't empty, fill `values` with the values of `state`, that way it'll preserve metadata passed by flags.
- Set `state.MetadataResult` to nil

This is needed because `MetadataSurvey` sets `state.MetadataResult` to the selected options [here](https://github.com/cli/cli/blob/9f84f0ffa1d5b3141399b7ed499eb81b69ef74d2/pkg/cmd/pr/shared/survey.go#L239), but if the user doesn't select anything or adds metadata then [`AddMetadataToIssueParams` fails the get the IDs](https://github.com/cli/cli/blob/9f84f0ffa1d5b3141399b7ed499eb81b69ef74d2/pkg/cmd/pr/shared/params.go#L61). By setting it to `nil` it'll be properly filled [here](https://github.com/cli/cli/blob/9f84f0ffa1d5b3141399b7ed499eb81b69ef74d2/pkg/cmd/pr/shared/params.go#L45).


Demo:

![gh-metadata](https://user-images.githubusercontent.com/6853656/100130915-40537b00-2e7b-11eb-9210-d5a583924b4f.gif)

cristiand391 committed:
Prompt: avoid resetting PR/issue metadata if no option is selected

cristiand391 committed:
don't reset previously added metadata

mislav commented on pkg/cmd/pr/shared/survey.go:
```diff
@@ -345,19 +345,34 @@ func MetadataSurvey(io *iostreams.IOStreams, client *api.Client, baseRepo ghrepo
+		}
 	}
 
+	state.MetadataResult = nil
```
At first I wasn't clear why you did this, but then I tested it out and noticed that it fails in `AddMetadataToIssueParams` without this line.

I'm not convinced of this workaround, though, because it clears the cache of what was just fetched from the server and causes a whole bunch of records to be fetched again. I realize that fixing this properly is non-trivial, and I don't require that you add it to your PR. I'm currently exploring alternatives.

mislav reviewed (commented):
Thank you for the fix; and good catch!

I noticed that there were not tests for this piece of logic. I do not fault you for not adding one, because it looks like making a test for this method is pretty hard. I will try to take over from this point. 🙇

mislav committed:
Ensure efficient resolving of `issue/pr create` metadata to GraphQL IDs

For metadata types chosen in interactive flow, we fetch all records from
the API in order to be able to display a multi-select interface.

For metadata defined via command-line flags, we resolve records that can
be looked up directly, avoiding fetching the entirety of expensive
datasets (e.g. all members of an organization) if we can.

The new approach ensures efficient fetching when interactive flow is
combined with values from flags.

mislav committed:
Make MetadataSurvey testable by accepting an interface

mislav committed:
AskStubber now throws a more descriptive error when stubs do not match

mislav committed:
Allow interactive `pr create` even if we failed to look up commits

mislav reviewed (approved):
I've pushed changes to resolve the metadata fetching conflict and to make `MetadataSurvey` easier to test, so we at least have some coverage around the interactive flow.

chenguru860530 reviewed (commented):
api/queries_repo.go

chenguru860530 commented:
api/queries_repo.go

pes2020kk reviewed (commented):
ios
